hawk/app/models/report.rb

```diff
@@ -1,69 +1,225 @@
   end
 
-  class << self
-    def parse(file)
-      hash = Digest::SHA1.hexdigest(
-        file.basename(".tar.bz2").to_s
-      )
+  def peinput_version(path)
+    nvpair = `CIB_file=#{path} cibadmin -Q --xpath "/cib/configuration//crm_config//nvpair[@name='dc-version']" 2>/dev/null`
+    m = nvpair.match(/value="([^"]+)"/)
+    return nil unless m
+    m[1]
+  end
+
+  def transition_cmd(hb_report, path, cmd)
+    source = archive
+    source = hb_report.path if File.directory?(hb_report.path)
+    Util.capture3("crm", "history", stdin_data: "source #{source}\n#{cmd}\n")
+  end
+
+  def info(hb_report, path)
+    out, err, status = transition_cmd hb_report, path, "transition #{path} nograph"
+    info = out + err
+    info.strip!
+    info = _("No details available") if info.empty?
+    info.insert(0, _("Error:") + "\n") unless status.exitstatus == 0
+    info
+  end
+
+  def tags(hb_report, path)
+    out, err, status = transition_cmd hb_report, path, "transition tags #{path}"
+    out.split
+  end
+
+  def logs(hb_report, path)
+    out, err, status = transition_cmd hb_report, path, "transition log #{path}"
+    info = out + err
+    info.strip!
+    info = _("No details available") if info.empty?
+    info.insert(0, _("Error:") + "\n") unless status.exitstatus == 0
+    info
+  end
+
+  # Returns [success, data|error]
+  def graph(hb_report, path, format=:svg)
+
+    tpath = Pathname.new(hb_report.path).join(path)
+
+    # Apparently we can't rely on the dot file existing in the hb_report, so we
+    # just use ptest to generate it.  Note that this will fail if hacluster doesn't
+    # have read access to the pengine files (although, this should be OK, because
+    # they're created by hacluster by default).
+    require "tempfile"
+    tmpfile = Tempfile.new("hawk_dot")
+    tmpfile.close
+    File.chmod(0666, tmpfile.path)
+    out, err, status = Util.run_as('hacluster', 'crm_simulate', '-x', tpath.to_s, format == :xml ? "-G" : "-D", tmpfile.path.to_s)
+    #status = Util.safe_x("/usr/sbin/crm_simulate", "-x", tpath, format == :xml ? "-G" : "-D", tmpfile.path)
+    rc = status.exitstatus
+
+    # TODO(must): handle failure of above
+
+    ret = [false, err]
+    if rc != 0
+      ret = [false, err]
+    elsif format == :xml || format == :json
+      # Can't use send_file here, server whines about file not existing(?!?)
+      # send_data File.new(tmpfile.path).read, :type => (params[:munge] == "txt" ? "text/plain" : "text/xml"), :disposition => "inline"
+      ret = [true, File.new(tmpfile.path).read]
+    else
+      svg, err, status = Util.capture3("/usr/bin/dot", "-Tsvg", tmpfile.path)
+      if status.exitstatus == 0
+        ret = [true, svg]
+      else
+        ret = [false, err]
+      end
+    end
+    tmpfile.unlink
+    ret
+  end
 
-      match = file.basename.to_s.match(
-        /(\d{4}-\d{2}-\d{2}_\d{2}:\d{2})/
-      )
+  # Returns the diff as a text or html string
+  def diff(hb_report, path, left, right, format=:html)
```

Comment:
Metrics/AbcSize: Assignment Branch Condition size for diff is too high. [18.79/15]<br>Style/SpaceAroundEqualsInParameterDefault: Surrounding space missing in default value assignment.